commands/database-push.js

```diff
@@ -1,60 +1,55 @@
 'use strict';
 
 var Command = require('../lib/command');
-var requireDatabaseAccess = require('../lib/requireDatabaseAccess');
-var request = require('request');
+var requireAccess = require('../lib/requireAccess');
 var api = require('../lib/api');
-var responseToError = require('../lib/responseToError');
 var FirebaseError = require('../lib/error');
 var RSVP = require('rsvp');
 var utils = require('../lib/utils');
-var querystring = require('querystring');
 var chalk = require('chalk');
 var logger = require('../lib/logger');
 var fs = require('fs');
 var Firebase = require('firebase');
 var _ = require('lodash');
+var readJSONInput = require('../lib/readJSONInput');
 
 module.exports = new Command('database:push <path> [infile]')
   .description('add a new JSON object to a list of data in your Firebase')
   .option('-d, --data <data>', 'specify escaped JSON directly')
-  .before(requireDatabaseAccess)
+  .before(requireAccess)
   .action(function(path, infile, options) {
     if (!_.startsWith(path, '/')) {
       return utils.reject('Path must begin with /', {exit: 1});
     }
 
-    return new RSVP.Promise(function(resolve, reject) {
-      var inStream = utils.stringToStream(options.data) || (infile ? fs.createReadStream(infile) : process.stdin);
+    var inStream = utils.stringToStream(options.data) || (infile ? fs.createReadStream(infile) : process.stdin);
+    var url = utils.addSubdomain(api.realtimeOrigin, options.instance) + path + '.json?';
 
-      var url = utils.addSubdomain(api.realtimeOrigin, options.instance) + path + '.json?';
-      var query = {auth: options.databaseAdminToken};
-
-      url += querystring.stringify(query);
+    if (!infile && !options.data) {
+      utils.explainStdin();
+    }
 
-      if (!infile && !options.data) {
-        utils.explainStdin();
+    return readJSONInput(inStream)
+    .then(function(input) {
+      return api.request('POST', url, {
+        auth: true,
+        data: input,
+        origin: ''
```

Comment:
Would prefer refactoring `origin` to be optional rather than passed in as empty string. Also, like with the `get` command above, the user may be piping in a 100MB file, so we don't actually want to parse the input but instead pipe it directly through. Like above, I think the best approach might be to keep the existing request more or less intact but pull in the access token.